### PR TITLE
fix: improve socket error handling

### LINKED ISSUE #383: Add Docker Socket availability check during connection creation
Currently, the `connect_with_` methods return a type of `Result<Docker, Error>` without performing any availability check. This means that even if the Docker Socket is not available, the method returns `Ok()` instead of `Err()`.

For example, when attempting to connect using `Docker::connect_with_socket_defaults()`, if `/var/run/docker.sock` is not available, it still returns `Ok()` instead of `Err()`.

To improve usability and prevent unexpected behavior, I propose adding an availability check during connection creation. This check should verify the availability of the Docker Socket and return an error if it's not accessible.

```rust
let docker = match Docker::connect_with_socket_defaults() {
  Ok(docker) => docker,
  Err(err) => {
    eprintln!("failed to connect to docker daemon: {:?}", err);
    return;
  }
};
```

### PR DESCRIPTION
**Pull request by KMJ-007**
I am new to rust, I was building docker tui, and came across the same issue as #383, i tried to fix this,

I don't know what other things I need to fix or add, this is going to be my 1st PR for rust, so if any issues need to be addressed please let me know,

closes #383 

had this question:
- should i modify existing examples or create new one for this?
- is there a need for a test for this?

thankyou for creating this lib, it is helping me a lot!!

### CODE REVIEW

**Comment by fussybeaver** [comment]
Thanks for this... could you also check if the socket is readable? I recall that some linux distributions require group access to the socket.

**Comment by KMJ-007** [comment]
> Thanks for this... could you also check if the socket is readable? I recall that some linux distributions require group access to the socket.

i have mac, in that it is working, not asking for any permission, don't have linux machine at the moment

image

**Comment by KMJ-007** [comment]
@fussybeaver , i got my friend's linux pc, and tested it

it is working, not asking for any permission

image

**Comment by KMJ-007** [comment]
it is running pop os

> @fussybeaver , i got my friend's linux pc, and tested it
> 
> it is working, not asking for any permission
> 
> image

**Comment by KMJ-007** [comment]
let me know what is show stopper for this PR, and what i can improve?

File: src/docker.rs [hunk 1]
```diff
@@ -4,6 +4,7 @@ use std::fs;
 use std::future::Future;
 #[cfg(feature = "ssl")]
 use std::io;
+use std::path::Path;
 #[cfg(feature = "ssl")]
 use std::path::{Path, PathBuf};
```
Comment (fussybeaver) [inline src/docker.rs:9]: The CI is failing because the `ssl` feature conflicts with this changeset

**Comment by KMJ-007** [comment]
All ci/cd's are green except docs one, i tried to check the logs but didn't understand why it is falling

**Comment by fussybeaver** [comment]
It looks like a doctest is failing, try adding a `,no_run` pragma to this line https://github.com/fussybeaver/bollard/blob/9435d91ef9d9d30cd95d85b3bcca9aba4b6a4e8e/src/grpc/driver/docker_container.rs#L121

**Comment by KMJ-007** [comment]
now something new is failing

**Comment by KMJ-007** [comment]
@fussybeaver  All Green

**Comment by mominul** [comment]
@fussybeaver Hi, can you kindly release a patch version of this crate with this change? I am facing this `HyperLegacyError`.
```
called `Result::unwrap()` on an `Err` value: Client(CreateContainer(HyperLegacyError { err: Error { kind: Connect, source: Some(Os { code: 2, kind: NotFound, message: "No such file or directory" }) } }))
```
 I am using `testcontainers_modules` crate which in turn uses `bollard` on macOS with `Docker version 26.1.1, build 4cf5afa`. Although docker is running I am getting this error. It might help me to troubleshoot if this change is available in the new release.

**Comment by mominul** [comment]
After patching the `bollard` dependency in `Cargo.toml`
```
[patch.crates-io]
bollard = { git = "https://github.com/fussybeaver/bollard.git" }
```
The error I get is:
```
Failed to start container: failed to initialize a docker client: Socket not found: /var/run/docker.sock
```

After enabling the `Allow privileged port mapping` option, now `bollard` is able to connect with docker successfully. :tada:

image

I solved this by following the lando/lando#3533 thread. But following the testcontainers/testcontainers-java#6045 and lando/lando#3533, in macOS docker desktop opens the `docker.sock` at `~/.docker/run/docker.sock` and `/var/run/docker.sock` is a symlink to it:
```
$ ll /var/run/docker.sock 
lrwxr-xr-x@ 1 root  daemon    38B Jul  9 23:55 /var/run/docker.sock@ -> /Users/mominul/.docker/run/docker.sock
```

So on macOS, `~/.docker/run/docker.sock` should be a fallback option, otherwise the users would need to face issues like I have faced.

Thank you for this crate! ❤️

**Pull request merged**